src/goto-programs/remove_const_function_pointers.cpp

```diff
@@ -0,0 +1,1013 @@
+
+ Purpose: Look up a symbol in the symbol table and return its value
+
+\*******************************************************************/
+
+exprt remove_const_function_pointerst::resolve_symbol(
+  const symbol_exprt &symbol_expr) const
+{
+  const symbolt &symbol=
+    symbol_table.lookup(symbol_expr.get_identifier());
+  return symbol.value;
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::try_resolve_function_call
+
+  Inputs:
+   expr - The expression to get the possible function calls
+   out_functions - The functions this expression could be resolved to
+
+ Outputs: Returns true if it was able to resolve the expression to some
+          specific functions. If this is the case, out_functions will contain
+          the possible functions.
+
+ Purpose: To resolve an expression to the specific function calls it can
+          be. This is different to try_resolve_expression which isn't
+          explicitly looking for functions and is instead just trying
+          to squash particular exprt structures.
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::try_resolve_function_call(
+  const exprt &expr, functionst &out_functions)
+{
+  const exprt &simplified_expr=simplify_expr(expr, ns);
+  bool resolved=false;
+  functionst resolved_functions;
+  if(simplified_expr.id()==ID_index)
+  {
+    const index_exprt &index_expr=to_index_expr(simplified_expr);
+    resolved=try_resolve_index_of_function_call(index_expr, resolved_functions);
+  }
+  else if(simplified_expr.id()==ID_member)
+  {
+    const member_exprt &member_expr=to_member_expr(simplified_expr);
+    resolved=try_resolve_member_function_call(member_expr, resolved_functions);
+  }
+  else if(simplified_expr.id()==ID_address_of)
+  {
+    address_of_exprt address_expr=to_address_of_expr(simplified_expr);
+    resolved=try_resolve_address_of_function_call(
+      address_expr, resolved_functions);
+  }
+  else if(simplified_expr.id()==ID_dereference)
+  {
+    const dereference_exprt &deref=to_dereference_expr(simplified_expr);
+    resolved=try_resolve_dereference_function_call(deref, resolved_functions);
+  }
+  else if(simplified_expr.id()==ID_typecast)
+  {
+    typecast_exprt typecast_expr=to_typecast_expr(simplified_expr);
+    resolved=
+      try_resolve_typecast_function_call(typecast_expr, resolved_functions);
+  }
+  else if(simplified_expr.id()==ID_symbol)
+  {
+    if(simplified_expr.type().id()==ID_code)
+    {
+      resolved_functions.insert(simplified_expr);
+      resolved=true;
+    }
+    else
+    {
+      LOG("Non const symbol wasn't squashed", simplified_expr);
+      resolved=false;
+    }
+  }
+  else
+  {
+    LOG("Unrecognised expression", simplified_expr);
+    resolved=false;
+  }
+
+  if(resolved)
+  {
+    out_functions.insert(resolved_functions.begin(), resolved_functions.end());
```

Comment:
Is `try_resolve_function_call` expected to be called with non-empty `out_functions` arguments?